Java/src/com/couchbase/litecore/fleece/MDict.java

```diff
@@ -13,92 +13,244 @@
+    public boolean remove(String key) {
+        return set(key, MValue.EMPTY);
+    }
+
+    public boolean set(String key, MValue value) {
+        if (key == null)
+            throw new IllegalArgumentException("The key cannot be null.");
 
-    private static native long init();
+        if (!isMutable())
+            throw new IllegalStateException("Cannot set items in a non-mutable MDict");
 
-    private static native void initInSlot(long handle, long mv, long parent);
+        MValue oValue = _map.get(key);
+        if (oValue != null) {
+            // Found in _map; update value:
+            if (value.isEmpty() && oValue.isEmpty())
+                return true; // no-op
+            mutate();
+            _count += (value.isEmpty() ? 0 : 1) - (oValue.isEmpty() ? 0 : 1);
+            _map.put(key, value);
+        } else {
+            // Not found; check _dict:
+            if (_dict != null && _dict.getSharedKey(key, getSharedKeys()) != null) {
+                if (value.isEmpty())
+                    _count--;
+            } else {
+                if (value.isEmpty())
+                    return true; // no-op
+                else
+                    _count++;
+            }
 
-    private static native void initAsCopyOf(long handle, long mDict, boolean isMutable);
+            mutate();
+            setInMap(key, value);
+        }
+        return true;
+    }
 
-    private static native long count(long handle);
+    /* Private Methods */
 
-    private static native boolean contains(long handle, String key);
+    private MValue setInMap(String key, MValue value) {
+        _newKey.add(key);
+        _map.put(key, value);
+        return value;
+    }
 
-    private static native long get(long handle, String key);
+    /* Encodable */
 
-    private static native boolean set(long handle, String key, long val);
+    @Override
+    public void encodeTo(Encoder enc) {
+        if (!isMutated()) {
+            if (_dict == null) {
+                enc.beginDict(0);
+                enc.endDict();
+            } else
+                enc.writeValue(_dict);
+        } else {
+            enc.beginDict(_count);
+            for (Map.Entry<String, MValue> entry : _map.entrySet()) {
+                MValue value = entry.getValue();
+                if (!value.isEmpty()) {
+                    enc.writeKey(entry.getKey());
+                    value.encodeTo(enc);
+                }
+            }
 
-    private static native boolean remove(long handle, String key);
+            if (_dict != null) {
+                FLDictIterator itr = new FLDictIterator();
+                try {
+                    itr.begin(_dict);
+                    String key;
+                    SharedKeys sk = new SharedKeys(getSharedKeys());
+                    while ((key = SharedKeys.getKey(itr, sk)) != null) {
+                        if (!_map.containsKey(key)) {
+                            enc.writeKey(key);
```

Comment:
Should not serialize if a `value` in the `_map` is `EMPTY`?

Reply:
It shouldn't. Why do you think it will serialize the empty value keys?

Reply:
same here. no issue.